File: Semester 4/Class11_Computer_Contest_Level_3_Notes_and_Homework_Nov_19/lvl3_hw_sol/split_a_sequence.java
import java.util.*;
import java.io.*;
public class homework {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st; 
    static int n, m, a[];
    public static void main(String[] args) throws IOException{
    	n = readInt(); m = readInt();  a = new int[n];
    	long lo = 0, hi = 0;
    	for(int i=0; i<n; i++) {
    		a[i] = readInt();
    		lo = Math.max(lo, a[i]); hi += a[i];
    	}
    	long ans = 0;
    	while(lo <= hi) {
    		long mid = (lo + hi)/2;
    		if(check(mid)) { ans = mid; hi = mid - 1;}
    		else lo = mid + 1;
    	}
    	System.out.println(ans);
    }
    static boolean check(long x) {
    	long sum = 0; int cnt = 1;
    	for(int i=0; i<n; i++) {
    		if(sum + a[i] > x) {
    			sum = a[i]; cnt ++;
    		}else {
    			sum += a[i];
    		}
    	}
    	return cnt <= m;
    }
    static String next () throws IOException {
        while (st == null || ! st.hasMoreTokens())
            st = new StringTokenizer(br.readLine().trim());
        return st.nextToken();
    }
    static long readLong () throws IOException {
        return Long.parseLong(next());
    }
    static int readInt () throws IOException {
        return Integer.parseInt(next());
    }
    static double readDouble () throws IOException {
        return Double.parseDouble(next());
    }
    static char readCharacter () throws IOException {
        return next().charAt(0);
    }  
    static String readLine () throws IOException {
        return br.readLine().trim();
    }
}

File: Semester 4/Class11_Computer_Contest_Level_3_Notes_and_Homework_Nov_19/lvl3_hw_sol/split_a_sequence.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
int N, K, a[100002]; ll ans, lo, hi;
bool check(ll x){
    int cnt = 1; ll sum = 0;
    for(int i=0; i<N; i++){
        if(sum + a[i] > x){ cnt++; sum = a[i]; }
        else sum += a[i];
    }
    return cnt <= K;
}
int main(){
    scanf("%d %d", &N, &K);
    for(int i=0; i<N; i++){
        scanf("%d", &a[i]); hi += a[i];  lo = max(lo, (ll)a[i]);
    }
    while(lo <= hi){
        ll mid = (lo + hi)/2;
        if(check(mid)){ ans = mid; hi = mid-1;}
        else lo = mid+1;
    }
    printf("%lld\n", ans);
}